kernel/src/sched/sched_class/fair.rs

```diff
@@ -0,0 +1,295 @@
+    ///    runnable threads (including the current running thread).
+    ///
+    /// The formula is chosen by 3 principles:
+    ///
+    /// 1. The scheduling period should reflect the running threads and CPUs;
+    /// 2. The scheduling period should not be too low to limit the overhead of context switching;
+    /// 3. The scheduling period should not be too high to ensure the scheduling latency
+    ///    & responsiveness.
+    fn period(&self) -> u64 {
+        let base_slice_clks = base_slice_clocks();
+        let min_period_clks = min_period_clocks();
+
+        let min_gran_clks = base_slice_clks * u64::from((1 + num_cpus()).ilog2());
+        // `+ 1` means including the current running thread.
+        (min_gran_clks * (self.threads.len() + 1) as u64).max(min_period_clks)
+    }
+
+    /// The normalized time slice for each thread in the runqueue, measured in vruntime clocks.
+    fn normalized_time_slice(&self) -> u64 {
+        self.period() / (self.threads.len() + 1) as u64
+    }
+
+    /// The time slice for each thread in the run queue, measured in sched clocks.
+    fn time_slice(&self, cur_weight: u64) -> u64 {
+        self.period() * cur_weight / (self.total_weight + cur_weight)
+    }
+
+    /// The minimum vruntime value for enqueued threads.
+    ///
+    /// This function is used to ensure the enqueued threads' vruntime be not far away from the
+    /// vruntime values in the run queue, causing possible starvation from them.
+    #[allow(unused)]
+    fn min_queued_vruntime(&self) -> u64 {
+        self.min_vruntime
+            .saturating_sub(self.normalized_time_slice())
+    }
+
+    fn update_min_vruntime(&mut self) {
+        self.min_vruntime = match self.threads.first() {
+            Some(thread) => thread.key(),
+            None => 0,
+        };
+    }
+
+    /// Pop a thread capable for running.
+    ///
+    /// If `target_cpu` is specified, this function filters out threads which
+    /// are not affined with it.
+    fn pop(&mut self, target_cpu: Option<CpuId>) -> Option<Arc<Thread>> {
+        let mut front = self.threads.lower_bound_mut(Bound::Unbounded);
+        let FairQueueItem(thread) = loop {
+            let thread = front.peek_next()?;
+            if target_cpu.is_none_or(|cpu| thread.0.atomic_cpu_affinity().load().contains(cpu)) {
+                let thread = front.remove_next().unwrap();
+                break thread;
+            }
+            front.next().unwrap();
+        };
+
+        self.total_weight -= thread.sched_attr().fair.weight.load(Relaxed);
+        self.update_min_vruntime();
+
+        Some(thread)
+    }
+}
+
+impl SchedClassRq for FairClassRq {
+    fn enqueue(&mut self, thread: Arc<Thread>) {
+        let fair_attr = &thread.sched_attr().fair;
+        // FIXME: Uncommenting this hangs the CI.
+        // fair_attr
+        //     .vruntime
+        //     .fetch_max(self.min_queued_vruntime(), SeqCst);
```

Comment:
The `min_queued_runtime` function is implemented as such to give some privilege to newly-initialized or long-waited threads. The subtractor is the normalized time slice of each thread in the run queue, whose meaning I think is described pretty clearly.

For the naming of `min_runtime`, the implementation of run queues per scheduling class doesn't include the current thread by design, so the names correspond to it.

Reply:
This is not how [Linux's CFS](https://github.com/torvalds/linux/blob/v2.6.24/kernel/sched_fair.c#L322-L325) maintains its `min_vruntime`.

```c
	/*
	 * maintain cfs_rq->min_vruntime to be a monotonic increasing
	 * value tracking the leftmost vruntime in the tree.
	 */
```

Our current `min_queued_vruntime` method does not return the _minimum_ vruntime of runnable tasks in a run queue, but the _second_ minimum vruntime _minus_ the nomalized time slice. This choice is really, really strange.

```rust
        // FIXME: Uncommenting this hangs the CI.
        // fair_attr
        //     .vruntime
        //     .fetch_max(self.min_queued_vruntime(), Relaxed);
```

We cannot take this value to update the vruntime of a newly-enqueued runnable task. Doing so may result the new task having the new minimum or the second minimum vruntime value in the run queue. This means that the `min_vruntime` of the runqueue may get decreased (the task's vruntime still increases due to `fetch_max`), breaking the monotonicity property.

In short, the `FairClassRq::min_vruntime` should keep the minimum vruntime of all runnable task, including the current one, and the value should be updated to ensure monotonicity.

In addition, the `vruntime` of a newly-enqueued task should be update to `rq.min_vruntime + penality` if `vurntime < rq.min_vruntime + penality`, where `penality` is a positive value controlled by the scheduler algorithm. See [the `place_entity` function](https://github.com/torvalds/linux/blob/49914084e797530d9baaf51df9eda77babc98fa8/kernel/sched_fair.c#L488).

Reply:
If you stick to including the current running task in the calculation of `min_vruntime`, then we need to move the updating code into `update_current`, which again causes the time complexity to be `O(log(n))`. This is because we cannot maintain the leftmost reference to the run queue together with it (self-referencing) and can only retrieve it from the `BTreeSet` every time. Or we could use `BinaryHeap`, but efficiency is challenged.

Also, although maintaining the monotonicity of vruntime solves the hanging of the CI (the `test_linux` one), I cannot figure out its necessity. Anyway, let the CI do the talking.

Reply:
> Also, although maintaining the monotonicity of vruntime solves the hanging of the CI (the test_linux one), ...

Great. This is very encouraging news.

> ... I cannot figure out its necessity. Anyway, let the CI do the talking.

Let me put it this way and see if it makes sense to you. How do you define "fair" in CFS _mathematically_? Apparently, you can't define "fair" as always scheduling the task with the least vruntime. This is what the scheduler _does_, not what the tasks _obtain_.

In my (humble) opinion, CFS being "fair" means holding the following invariant 

$|T_i - T_j| <= P / N$

is held for any $1 \le i < j \le N$ at any point of time. Here, $T_i$ ($T_j$) denotes the vruntime of the $i$-th ($j$-th) task in a local run queue, $N$ the length of the run queue, and $P$ is the time of a scheduling period. A scheduling period is defined as a period of time during which each runnable task is scheduled _at least once_. I believe this invariant captures the intuition of being fair.

Put it another way, we have the following invariant:

$T_{max} - T_{min} <= P / N$

where $T_{min}$ and $T_{max}$ are the minimum and maximum vruntimes of tasks in the run queue. 

While the two invariants sound good in theory, one technical difficulty in practice is that tasks are not always runnable, making their vruntimes meaningless. For example, an I/O bound task may be blocked on an event for an indefinite amount of time. When it wakes up and joins the run queue again, its vruntime may be much smaller than the vruntimes of the ones in the queue. We cannot use its vruntime as is because it is not the scheduler's fault that the I/O bound task has a low runtime. So we must assign it a new value so that its vruntime can once again serve as an effective indicator of the CPU time consumed by the task. And when doing so, we must keep the aforementioned invariant. Assuming that the $T'$ is the updated vruntime of the newly-enqueued task, we must have

$T_{min} \le T' \le T_{max}$,

otherwise, we break the aforementioned two invariants once the task is added to the run queue.

Since $T' >= T_{min}$,  we know $T_{min}$ must be a _monotonic increasing value_.

Furthermore, since $T' >= T_{min}$, we can express it as

$T' = T_{min} + T_{penality}$

where $T_{penality} >= 0$ is a value chosen by the scheduler using some complicated heuristics.

The reasoning above is why I suggest maintaining the minimum vruntime (as Linux does) of the runqueue. 

> If you stick to including the current running task in the calculation of min_vruntime, then we need to move the updating code into update_current, which again causes the time complexity to be O(log(n))

What I insist are the invariants, not a particular implementation. Keeping the invariants does not dictate the $O(\log(N))$ complexity.

Reply:
> In my (humble) opinion, CFS being "fair" means holding the following invariant
> 
> | T i − T j | <= P / N

IMHO instead, fairness means every task gains indistingushable amount of vruntime in the same scheduling period. In other words:

$| \Delta T_{i} - P/N | < \epsilon$

Vruntimes are not necessarily equal, but their **differentials** should. The comparison of vruntimes themselves cannot determine the time slice they gain from each period. Hence, the absolute value of vruntime only determine the executing order in each period, and should consider not only weight of tasks, but also their requirements of responsiveness and other properties.

> What I insist are the invariants, not a particular implementation. Keeping the invariants does not dictate the O ( log ⁡ ( N ) ) complexity.

If you take the current task into account, then `update_current` should update `min_vruntime` as well:

```rust
let (vruntime, weight) = attr.fair.update(...);
self.min_vruntime = vruntime;
```

However, the code above is not correct if the current updated vruntime is greater then the minimal vruntime of ready-for-run tasks. So we should do this instead:

```rust
self.min_vruntime = self.threads.first()
    .map_or(vruntime, |item| vruntime.min(item.key()));
```

That is O(log N) in implementation if no additional structures are involved. There are alternatives though:

1. Use `BinaryHeap` instead, as accessing the top don't introduce `O(log n)` complexity, however the overall efficiency is challenged compared to B-trees.

2. Cloning the first item of `BTreeMap` and maintains it in the struct. This approach increases the memory overhead and adds logic maintained manually by developers, if it is considerable.

Note that keeping the current task in the B-tree or other ordered collections is ill-formed and causes unexpected behavior since its key (i.e. vruntime) should not be modified when the item is in the collection, which is an invariant that they depend on.

Reply:
Since the argument about theories cannot solve the current problems in the CI, I suggest we both postpone it until the issues are solved.

Reply:
Consider using [`RBTree`](https://docs.rs/intrusive-collections/latest/intrusive_collections/rbtree/struct.RBTree.html)?

Reply:
> its key (i.e. vruntime) should not be modified when the item is in the collection, which is an invariant they depend on.

No, we can't. We can EITHER store `current` in the collection OR update its vruntime out of the collection — no other choice. It's simply because items in ordered collections are **ORDERED**. This variant cannot be violated.